### ISSUE TITLE
Triclinic distance bug

### ISSUE THREAD

**Issue by rmcgibbo** [Amber-MD/cpptraj#223]
I think I found a bug in the distance computation for some very skewed unit cells.

To reproduce, consider the following structure:
 ```
$ cat structure.pdb
CRYST1   20.000   20.000   20.000  41.55  56.39  17.52 P 1           1
MODEL        0
ATOM      1  O   HOH A   1       8.029   5.236   5.067  1.00  0.00          O
ATOM      2  H1  HOH A   1      13.771   5.666  14.130  1.00  0.00          H
TER       3      HOH A   1
ENDMDL
END
```

`cpptraj` gives a distance of 7.59A between these two atoms.

```
$ echo 'distance @1 @2 out distance.txt' | cpptraj  -y structure.pdb -p structure.pdb > /dev/null; cat distance.txt
#Frame      Dis_00001
       1       7.5932
```

However, I think I can find a closer distance between the two atoms. Here's my code:
```
import numpy as np
import scipy.spatial

# unit cell vectors
a = np.array([20., 0., 0.])
b = np.array([19.0720403, 6.0214001, 0.])
c = np.array([11.0718369, 14.6431303, 7.9368222])

# coordinates
x0 = np.array([8.029, 5.236, 5.067])
x1 = np.array([13.771, 5.666, 14.13])

def angle(v1, v2):
    return np.rad2deg(np.arccos(np.dot(v1, v2) / (np.linalg.norm(v1) *
                                                  np.linalg.norm(v2))))

# confirm that these unit vectors match the unit cell lengths and angles
# in the PDB file
print('alpha angle ', angle(b, c))
print('beta angle  ', angle(a, c))
print('gamma angle ', angle(a, b))
print('lengths     ', [np.linalg.norm(v) for v in (a, b, c)])
print()

offset_1 = (-2 * a) + (2 * b) + (-1 * c)
print('distance (-2,2,-1)  ',
      scipy.spatial.distance.euclidean(x0, x1 + offset_1))

offset_2 = (-4 * a) + (5 * b) + (-2 * c)
print('distance (-4,5,-2)  ',
      scipy.spatial.distance.euclidean(x0, x1 + offset_2))
```

Running this, I see that using `(-2,2,-1)` translations of the unit cells gives the distance reported by cpptraj, but using `(-4,5,-2)` translations gives a shorter distance.
```
$ python wut.py
alpha angle  41.5534977768
beta angle   56.38622
gamma angle  17.5218818781
lengths      [20.0, 19.9999995092276, 19.999999598934806]

distance (-2,2,-1)   7.590369080826602
distance (-4,5,-2)   7.002420438132136
```

**Comment by swails** [comment]
Are you experienced enough with GROMACS to know how well they handle this?

**Comment by drroe** [comment]
I'm looking into this now, thanks for the report.

(As an aside, this is officially the weirdest unit cell I've *ever* seen!)

**Comment by drroe** [comment]
OK, I can confirm the results of @rmcgibbo, illustrated in this figure:
![image](https://cloud.githubusercontent.com/assets/10144183/12203399/0690ed56-b5ec-11e5-986f-73b715f5b0d6.png)
I used cpptraj `vector ucell` commands to generate the unit cell vectors (red, green, and blue licorice), `replicatecell` to generate the imaged structures, then calculated the distances with VMD. The original atoms are in grey, the cpptraj (-2, 2, -1) atoms are in cyan, and the -4, 5, -2 atoms are in orange.

I believe the issue is that cpptraj assumes that once two atoms are wrapped back into the main unit cell (i.e. 0,0,0), the minimum imaged distance from atom0 to atom1 is either atom0 to atom1 in the main cell or can be found from atom0 only one cell over in any direction to atom1 in the main cell. However, that is not the case here because the cell is so narrow. I've confirmed this by modifying the search to go a few cells over in every direction:
```
[droe@adenine src]$ ./a.out 
Volume= 954.271
c1(frac)={ 0.701866 -0.686169 0.639384 }  c2(frac)={ 2.94153 -3.39755 1.78301 }
c1(frac0)={ 0.701866 0.313831 0.639384 }  c2(frac0)={ 0.941532 0.602452 0.783006 }
c1(0)={ 27.1012 11.2568 5.067}  c2(0)={ 38.9892 15.0986 6.20518 }
Scanning images from -5 to 5
Min(2, -2, 1)= 6.97543
```
The simple solution of course would be to expand the search by default; however this will severely impact performance (125 distance calcs vs 27 distance calcs for an increase of just 1 unit cell in each direction) so I'm not keen on doing that. There are more clever ways to find the minimum imaged distance but given that this can probably be considered a corner case for most biological systems I think this can be given low priority for now. It's something I would like to fix soon though. If anyone wants to take a crack at it I wouldn't mind :-).

**Comment by rmcgibbo** [comment]
>  (As an aside, this is officially the weirdest unit cell I've ever seen!)

It was just randomly generated :)

> Are you experienced enough with GROMACS to know how well they handle this?

Something like this
```
$ cat index.ndx
[A]
1  2
$ gmx distance -f structure.pdb -n index.ndx -select A

Reading frame       0 time    0.000   Warning: Triclinic box is too skewed.
         Box (3x3):
            Box[    0]={ 2.00000e+00,  0.00000e+00,  0.00000e+00}
            Box[    1]={ 1.90722e+00,  6.02077e-01,  0.00000e+00}
            Box[    2]={ 1.10707e+00,  1.46505e+00,  7.92482e-01}
         Can not fix pbc.
 '', 2 atoms
Last frame          0 time    0.000
Analyzed 1 frames, last time 0.000
A:
  Number of samples:  1
  Average distance:   1.074  nm
  Standard deviation: 0.000  nm
```

That is, giving a warning and then returning the wrong result. The warning / error seems like a good idea.

**Comment by rmcgibbo** [comment]
My intuition really is that it should be possible to calculate the distance without needing to "search" manually over the 26 (or 63...) translations at all. At least, it should be possible to sufficiently characterize the conditions that the box vectors must satisfy such that no search is necessary, such that it can be avoided in the general case.

**Comment by swails** [comment]
As I understand it, the formula in Tuckerman's text is valid as long as the distance between the centers is closer than the nearest unit cell edge.  Which is why that's all you need to do for MD.

I feel like we should at the very least be able to calculate how many images we need to calculate over.  But in 2 dimensions, you can demonstrate that, as the angle between the two unit cell vectors approaches 0, the number of unit cell translations needed to get the "minimum" image of the center with one of the corners goes arbitrarily high (I think, this is based on what I'm visualizing in my head right now, but it makes sense).

I agree it seems we should be able to find a closed-form expression to calculate the maximum number of displacements we need to test, but I just don't know that there is an efficient way to determine *which* image has the minimum distance rigorously for an arbitrary shape without testing all of those displacements explicitly -- especially not in 3 dimensions.

**Comment by swails** [comment]
FTR -- [here is the GROMACS code for triggering the skewed-box warning](http://redmine.gromacs.org/projects/gromacs/repository/revisions/master/entry/src/gromacs/pbcutil/pbc.cpp#L154)

Seems easy enough to adopt this in cpptraj and MDTraj

**Comment by drroe** [comment]
Addressed by #239
